From f7284ebf762aae9d49dcad570b45470ea0c81cad Mon Sep 17 00:00:00 2001
From: Jarkko Nikula <jarkko.nikula@linux.intel.com>
Date: Wed, 18 Sep 2019 13:48:47 +0300
Subject: [PATCH 5/8] spi: dw-pci: WIP. Add Elkhart Lake PSE DMA support

WIP. Only build tested due HW and simulation environment availability and
readiness.

Signed-off-by: Jarkko Nikula <jarkko.nikula@linux.intel.com>
---
 drivers/spi/spi-dw-mid.c | 36 +++++++++++++++++++++++++++++++++++-
 drivers/spi/spi-dw-pci.c |  1 +
 2 files changed, 36 insertions(+), 1 deletion(-)

diff --git a/drivers/spi/spi-dw-mid.c b/drivers/spi/spi-dw-mid.c
index 2663bb1..b3cca64 100644
--- a/drivers/spi/spi-dw-mid.c
+++ b/drivers/spi/spi-dw-mid.c
@@ -35,7 +35,7 @@ static bool mid_spi_dma_chan_filter(struct dma_chan *chan, void *param)
 	return true;
 }
 
-static int mid_spi_dma_init(struct dw_spi *dws)
+static int mid_spi_dma_init_pci(struct dw_spi *dws)
 {
 	struct pci_dev *dma_dev;
 	struct dw_dma_slave *tx = dws->dma_tx;
@@ -76,6 +76,40 @@ static int mid_spi_dma_init(struct dw_spi *dws)
 	return -EBUSY;
 }
 
+static int mid_spi_dma_init_generic(struct dw_spi *dws)
+{
+	struct device *dev = &dws->master->dev;
+	dma_cap_mask_t mask;
+
+	dma_cap_zero(mask);
+	dma_cap_set(DMA_SLAVE, mask);
+
+	dws->rxchan = dma_request_slave_channel_compat(mask, NULL, NULL,
+						       dev, "rx");
+	if (dws->rxchan)
+		return -ENODEV;
+	dws->master->dma_rx = dws->rxchan;
+
+	dws->txchan = dma_request_slave_channel_compat(mask, NULL, NULL,
+						       dev, "tx");
+	if (dws->txchan) {
+		dma_release_channel(dws->rxchan);
+		return -ENODEV;
+	}
+	dws->master->dma_tx = dws->txchan;
+
+	dws->dma_inited = 1;
+	return 0;
+}
+
+static int mid_spi_dma_init(struct dw_spi *dws)
+{
+	if (mid_spi_dma_init_generic(dws))
+		return mid_spi_dma_init_pci(dws);
+
+	return 0;
+}
+
 static void mid_spi_dma_exit(struct dw_spi *dws)
 {
 	if (!dws->dma_inited)
diff --git a/drivers/spi/spi-dw-pci.c b/drivers/spi/spi-dw-pci.c
index 528d4d6..618667f 100644
--- a/drivers/spi/spi-dw-pci.c
+++ b/drivers/spi/spi-dw-pci.c
@@ -45,6 +45,7 @@ static struct spi_pci_desc spi_pci_mid_desc_2 = {
 };
 
 static struct spi_pci_desc spi_pci_ehl_desc = {
+	.setup = dw_spi_mid_init,
 	.num_cs = 2,
 	.bus_num = -1,
 	.max_freq = 100000000,
-- 
2.7.4

